### ISSUE TITLE
Running Godot Integration tests in another crate

### ISSUE THREAD

**Issue by StatisMike** [godot-rust/gdext#480]
Is it possible (and if so, to what extent) to use current Godot integration tests, as implemented in `gdext` in another crate?

I think that many crates that are `gdext` extensions would greatly benefit from such documentation or the addition of that kind of exported functionality.

**Comment by Bromeon** [comment]
> Is it possible (and if so, to what extent) to use current Godot integration tests, as implemented in `gdext` in another crate?

Currently not, but I agree it might be nice, maybe in its own crate `godot-test`.

**Comment by StatisMike** [comment]
I will try to set up integration tests in my crate on basis of what is implemented in `gdext`. Will share experiences of what is needed at minimum to export this functionality.

**Comment by Bromeon** [comment]
Sounds good. If you get to a point where a lot of effort is needed to proceed, don't hesitate to discuss the design 🙂 

An important point for me is that we keep the scope of such a framework limited, and not open another long-lasting construction site right now. Ideally, features would match more or less exactly the ones used internally for gdext. A highly configurable test runner is not something I plan on maintaining at the current stage of the project; there are too many other important features in the pipeline 😉

**Comment by StatisMike** [comment]
Ideally I would like it to keep it identical to the one used by `itest`, though it isn't clear to me how to handle licensing in this if it should be a separate crate. Will certainly discuss the design if I get it running :)

**Comment by StatisMike** [comment]
@Bromeon Had it running outside of the crate, in the process of testing the rust-side tests and benchmarks, and rewriting whole `GDScriptTestRunner` on the Rust side to not force users to copy-paste Godot-side code.

At the current point in time:

- I think it won't be possible (or really hacky and risky) to create a standalone integration test crate that won't make it necessary to create a Godot project to run things inside. The minimal requirement will be for the user to create _gdextension_ and scene with ported `GDScriptTestRunner`, probably using shell script to run the Godot. I've saw two scenarios that will differ in a way Godot need to be started:
    1. if the user is writing tests for _gdextension_ that already has a Godot project linked to it (so: writing code for his Godot application), they will need to create empty _test scene_ for the integration tests to run in. Godot allows running specific scene from the command line, so they will just need to run the shell script argument pointing to created _test scene_.
    2. if the user is writing tests for a standalone crate providing an extension to `gdext` itself, they will need to make that crate a workspace crate and create similiar `crate + godot` construct as `gdext`'s `itest`.
Both routes I wish to document thoroughly in documentation.
- Above said, I'm thinking about the possiblity of running the integration tests in the editor. What were the concerns that made it not viable in `gdext`?
- I'm keeping the macros very similar to the ones in `itest`: `#[gdidest]` and `#[gdbench]` currently, possibly a rename later. Wanted to make it more specific than `#[itest]` and `#[bench]`, as they seem too generic if `godot-test` were to be used in scenario 1. I didn't reimplement `KvParser` in its entirety, though current parsing will need some polishing, as it is now implemented as MVP.
- I'm not publishing the code ATM, as much of it is borrowed from `gdext`. Will need a discussion about handling it - I don't want to reimplement what was already created and is working nicely and take credit for just exposing it in another way. Git history shows me that most - if not all of it was - created by you, so I would love a chat with you about exposing it and publishing in the future.
